File: frontend/src/services/industryCollaborationService.ts
import { api } from "./api";

export type IndustryCollaborationStatus =
  | "DRAFT"
  | "SUBMITTED"
  | "UNDER_REVIEW"
  | "MODIFICATION_REQUESTED"
  | "ACCEPTED"
  | "REJECTED"
  | "WITHDRAWN";

export interface UniversityProposal {
  id: number;
  rfp_id: number;
  invitation_id: number;
  university_id: number;
  submitted_by: number;

  title: string;
  solution: string;
  technical_approach: string;

  research_methodology: string | null;
  required_resources: string | null;
  estimated_cost: number | null;
  expected_timeline_days: number | null;
  faculty_team: string | null;
  expected_outcomes: string | null;
  technology_requirements: string | null;

  status:
    | "DRAFT"
    | "SUBMITTED"
    | "UNDER_EVALUATION"
    | "SHORTLISTED"
    | "REJECTED"
    | "WITHDRAWN";

  submitted_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface IndustryCollaboration {
  id: number;

  university_proposal_id: number;
  industry_id: number;
  submitted_by: number;

  title: string;
  collaboration_description: string;

  funding_amount: number | null;

  technical_mentorship: string | null;
  industry_experts: string | null;
  infrastructure_resources: string | null;
  technology_support: string | null;
  internship_support: string | null;
  pilot_deployment_support: string | null;
  commercialization_support: string | null;

  proposed_duration_days: number | null;
  response_deadline: string | null;
  additional_terms: string | null;

  status: IndustryCollaborationStatus;

  submitted_at: string | null;
  reviewed_at: string | null;

  created_at: string;
  updated_at: string;
}

export interface CreateIndustryCollaborationRequest {
  university_proposal_id: number;

  title: string;
  collaboration_description: string;

  funding_amount?: number;

  technical_mentorship?: string;
  industry_experts?: string;
  infrastructure_resources?: string;
  technology_support?: string;
  internship_support?: string;
  pilot_deployment_support?: string;
  commercialization_support?: string;

  proposed_duration_days?: number;
  response_deadline?: string;

  additional_terms?: string;
}

export interface IndustryCollaborationDecisionRequest {
  decision:
    | "ACCEPTED"
    | "REJECTED"
    | "MODIFICATION_REQUESTED";

  remarks?: string;
}

/* ============================================================
   INDUSTRY: SHORTLISTED PROPOSALS
============================================================ */

export async function getShortlistedProposals(): Promise<
  UniversityProposal[]
> {
  const response = await api.get<UniversityProposal[]>(
    "/university-proposals/shortlisted"
  );

  return response.data;
}

/* ============================================================
   INDUSTRY: SUBMIT COLLABORATION
============================================================ */

export async function createIndustryCollaboration(
  payload: CreateIndustryCollaborationRequest
): Promise<IndustryCollaboration> {
  const response =
    await api.post<IndustryCollaboration>(
      "/industry-collaboration",
      payload
    );

  return response.data;
}

/* ============================================================
   INDUSTRY: MY COLLABORATIONS
============================================================ */

export async function getMyIndustryCollaborations(): Promise<
  IndustryCollaboration[]
> {
  const response =
    await api.get<IndustryCollaboration[]>(
      "/industry-collaboration/my"
    );

  return response.data;
}

/* ============================================================
   UNIVERSITY: RECEIVED COLLABORATIONS
============================================================ */

export async function getUniversityCollaborations(): Promise<
  IndustryCollaboration[]
> {
  const response =
    await api.get<IndustryCollaboration[]>(
      "/industry-collaboration/university"
    );

  return response.data;
}

/* ============================================================
   UNIVERSITY: DECISION
============================================================ */

export async function decideIndustryCollaboration(
  collaborationId: number,
  payload: IndustryCollaborationDecisionRequest
): Promise<IndustryCollaboration> {
  const response =
    await api.post<IndustryCollaboration>(
      `/industry-collaboration/${collaborationId}/decision`,
      payload
    );

  return response.data;
}

export async function getAcceptedGovernmentCollaborations(): Promise<
  IndustryCollaboration[]
> {
  const response =
    await api.get<IndustryCollaboration[]>(
      "/industry-collaboration/government/accepted"
    );

  return response.data;
}

File: mediator_backend/app/services/industry_collaboration_service.py
from datetime import datetime, timezone

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.models.industry_collaboration import (
    IndustryCollaborationProposal,
    IndustryCollaborationStatus,
)
from app.models.organization import (
    Organization,
    OrganizationType,
)
from app.models.university_proposal import (
    UniversityProposal,
    UniversityProposalStatus,
)
from app.models.user import User, UserRole
from app.schemas.industry_collaboration import (
    IndustryCollaborationCreate,
)


def submit_collaboration_proposal(
    db: Session,
    collaboration_data: IndustryCollaborationCreate,
    current_user: User,
) -> IndustryCollaborationProposal:

    # ---------------------------------------------------------
    # 1. Validate user role
    # ---------------------------------------------------------
    if current_user.role not in {
        UserRole.INDUSTRY_ADMIN,
        UserRole.INDUSTRY_MEMBER,
    }:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only industry users can submit collaboration proposals.",
        )

    # ---------------------------------------------------------
    # 2. Validate organization association
    # ---------------------------------------------------------
    if not current_user.organization_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Industry user is not associated with an organization.",
        )

    # ---------------------------------------------------------
    # 3. Load and validate industry organization
    # ---------------------------------------------------------
    industry = (
        db.query(Organization)
        .filter(
            Organization.id == current_user.organization_id
        )
        .first()
    )

    if not industry:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Industry organization not found.",
        )

    if industry.organization_type != OrganizationType.INDUSTRY:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only industry organizations can submit collaboration proposals.",
        )

    if not industry.is_active:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="This industry organization is inactive.",
        )

    # ---------------------------------------------------------
    # 4. Load university proposal
    # ---------------------------------------------------------
    university_proposal = (
        db.query(UniversityProposal)
        .filter(
            UniversityProposal.id
            == collaboration_data.university_proposal_id
        )
        .first()
    )

    if not university_proposal:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="University proposal not found.",
        )

    # ---------------------------------------------------------
    # 5. Industry collaboration only for shortlisted proposals
    # ---------------------------------------------------------
    if (
        university_proposal.status
        != UniversityProposalStatus.SHORTLISTED
    ):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=(
                "Industry collaboration is available only "
                "for shortlisted university proposals."
            ),
        )

    # ---------------------------------------------------------
    # 6. Prevent duplicate proposal from same industry
    # ---------------------------------------------------------
    existing = (
        db.query(IndustryCollaborationProposal)
        .filter(
            IndustryCollaborationProposal.university_proposal_id
            == university_proposal.id,
            IndustryCollaborationProposal.industry_id
            == industry.id,
        )
        .first()
    )

    if existing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=(
                "This industry has already submitted a collaboration "
                "proposal for this university proposal."
            ),
        )

    # ---------------------------------------------------------
    # 7. Create collaboration proposal
    # ---------------------------------------------------------
    collaboration = IndustryCollaborationProposal(
        university_proposal_id=university_proposal.id,
        industry_id=industry.id,
        submitted_by=current_user.id,

        title=collaboration_data.title,
        collaboration_description=(
            collaboration_data.collaboration_description
        ),

        funding_amount=collaboration_data.funding_amount,

        technical_mentorship=(
            collaboration_data.technical_mentorship
        ),
        industry_experts=collaboration_data.industry_experts,
        infrastructure_resources=(
            collaboration_data.infrastructure_resources
        ),
        technology_support=(
            collaboration_data.technology_support
        ),
        internship_support=(
            collaboration_data.internship_support
        ),
        pilot_deployment_support=(
            collaboration_data.pilot_deployment_support
        ),
        commercialization_support=(
            collaboration_data.commercialization_support
        ),

        proposed_duration_days=(
            collaboration_data.proposed_duration_days
        ),
        
        response_deadline=(
            collaboration_data.response_deadline
        ),

        additional_terms=(
            collaboration_data.additional_terms
        ),

        status=IndustryCollaborationStatus.SUBMITTED,
        submitted_at=datetime.now(timezone.utc),
    )

    # ---------------------------------------------------------
    # 8. Save
    # ---------------------------------------------------------
    db.add(collaboration)
    db.commit()
    db.refresh(collaboration)

    return collaboration


def decide_collaboration_proposal(
    db: Session,
    collaboration_id: int,
    decision: str,
    remarks: str | None,
    current_user: User,
) -> IndustryCollaborationProposal:

    # ---------------------------------------------------------
    # 1. Validate university role
    # ---------------------------------------------------------
    if current_user.role not in {
        UserRole.UNIVERSITY_ADMIN,
        UserRole.FACULTY,
    }:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=(
                "Only university users can review "
                "industry collaboration proposals."
            ),
        )

    # ---------------------------------------------------------
    # 2. Load collaboration proposal
    # ---------------------------------------------------------
    collaboration = (
        db.query(IndustryCollaborationProposal)
        .filter(
            IndustryCollaborationProposal.id == collaboration_id
        )
        .first()
    )

    if not collaboration:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Industry collaboration proposal not found.",
        )

    # ---------------------------------------------------------
    # 3. Verify associated university proposal
    # ---------------------------------------------------------
    university_proposal = collaboration.university_proposal

    if not university_proposal:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Associated university proposal not found.",
        )

    # ---------------------------------------------------------
    # 4. Verify university ownership
    # ---------------------------------------------------------
    if current_user.organization_id != university_proposal.university_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=(
                "You are not authorized to review "
                "this collaboration proposal."
            ),
        )

    # ---------------------------------------------------------
    # 5. Validate decision
    # ---------------------------------------------------------
    allowed_decisions = {
        IndustryCollaborationStatus.ACCEPTED.value,
        IndustryCollaborationStatus.REJECTED.value,
        IndustryCollaborationStatus.MODIFICATION_REQUESTED.value,
    }

    if decision not in allowed_decisions:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=(
                "Decision must be ACCEPTED, REJECTED, "
                "or MODIFICATION_REQUESTED."
            ),
        )

    # ---------------------------------------------------------
    # 6. Validate current state
    # ---------------------------------------------------------
    allowed_current_statuses = {
        IndustryCollaborationStatus.SUBMITTED,
        IndustryCollaborationStatus.UNDER_REVIEW,
        IndustryCollaborationStatus.MODIFICATION_REQUESTED,
    }

    if collaboration.status not in allowed_current_statuses:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=(
                "This collaboration proposal cannot be reviewed "
                "in its current state."
            ),
        )

    # ---------------------------------------------------------
    # 7. Apply university decision
    # ---------------------------------------------------------
    collaboration.status = IndustryCollaborationStatus(decision)
    collaboration.reviewed_at = datetime.now(timezone.utc)

    # ---------------------------------------------------------
    # 8. Store review remarks
    # ---------------------------------------------------------
    if remarks:
        existing_terms = collaboration.additional_terms or ""

        collaboration.additional_terms = (
            f"{existing_terms}\n\n"
            f"University Review: {remarks}"
        ).strip()

    # ---------------------------------------------------------
    # 9. Save
    # ---------------------------------------------------------
    db.commit()
    db.refresh(collaboration)

    return collaboration